### PR TITLE
📌 Pin Clippy to a nightly 📌

### PR DESCRIPTION
**Pull request by ebroto**
changelog: Pin Clippy to a specific nightly version (No more master/custom toolchain required to compile Clippy)

Addresses partially #5561. As proposed there in [this comment](https://github.com/rust-lang/rust-clippy/issues/5561#issuecomment-623109095), this kicks off the process, to help us get acquainted with how the syncs should work, before working on improving the tooling.

Open questions:
* When performing a rustup, we will need to exclude the commits that were merged that same day, or else wait until that nightly is released. I did not update the documentation about this part, mainly because I'm not sure about how to do that.
* When should we perform the rustups now? My first idea is to do it at the same time we do the clippyups, to have a clear cadence and to avoid the two copies of the repo to diverge enough to make the process painful.
* Who does the rustups now? If we follow my previous idea and do both rustup and clippyup at the same time, it would be more work for @flip1995 who currently does the clippyups. I would prefer to establish some kind of rotation to spead the work. Other ideas?
* I'm not sure if this affects the release process in any way.
* ???

@rust-lang/clippy thoughts?

r? @flip1995

### CODE REVIEW

**Comment by ebroto** [comment]
(converting to draft as I want to test the changes manually in the rustc repo)

**Comment by matthiaskrgr** [comment]
>When performing a rustup, we will need to exclude the commits that were merged that same day, or else wait until that nightly is released.

Can we specify a commit up to that we sync when pulling from `rustc` into `rust-lang/clippy`?
If so, maybe we could have some kind of script which finds out the latest version of rustc nightly (perhaps via `rustup check`?) and performs a sync up to that rust-nightly commit, and then edits the `rust-toolchain` file accordingly?

**Comment by flip1995** [comment]
> * When performing a rustup, we will need to exclude the commits that were merged that same day, or else wait until that nightly is released. I did not update the documentation about this part, mainly because I'm not sure about how to do that.

I like @matthiaskrgr idea. So the process of a Rustup would be to check the commit of the latest nightly, then checkout this commit in the Rust repo and do the sync from the detached head. 

We can build tooling for this later. There is also [`gitree`](https://crates.io/crates/gitree), which @mominul implemented and is still waiting on my review. I suggest to enhance this tool and use it.

> * When should we perform the rustups now? My first idea is to do it at the same time we do the clippyups, to have a clear cadence and to avoid the two copies of the repo to diverge enough to make the process painful.

Yeah, we should do it together with (right after) the rustups. Rustup PRs should get `p=1` in the rust repo, so that the process is faster. If the process should get painful, we can do those rustups every week, instead of every other week.

> * Who does the rustups now?

I don't mind doing them, it's really not that much work. I'll post on Zulip if I should not get to do them for some reason. 

> * I'm not sure if this affects the release process in any way.

I don't think so? 🤔 We'll see, I guess

File: CONTRIBUTING.md
Comment (flip1995) [inline CONTRIBUTING.md:182]: Whooo 🎉

File: CONTRIBUTING.md
Comment (flip1995) [inline CONTRIBUTING.md:191]: If we should decide to do them together with the Clippyups, we should mention that cycle. Something like

"we have this and that cycle. If you think that a out-of-cycle sync should be done, do this and that."

File: setup-toolchain.sh
Comment (flip1995) [inline setup-toolchain.sh:1]: RIP setup-toolchain script 2019-2020 😢

**Review by flip1995 — COMMENTED** [review]
I'll have to do in depth review and think about this for a bit longer, so that we don't forget about something. Thanks for moving this forward ❤️

**Comment by matthiaskrgr** [comment]
Right now, the `rustfmt` component is missing from nightly and it seems that when building clippy with a nightly that does not offer a required (in the `rust-toolchain` file) component, rustup simply skips the component.
````
info: syncing channel updates for 'nightly-2020-11-29-x86_64-unknown-linux-gnu'
info: latest update on 2020-11-29, rust version 1.50.0-nightly (e37f25aa3 2020-11-28)
warning: Force-skipping unavailable component 'rustfmt-x86_64-unknown-linux-gnu'
info: downloading component 'cargo'
...
````

This might be something to be aware of, I hope it won't break our neck later at some point....

It also installs clippy from that nightly for some reason which seems a bit excessive? :smile:

**Comment by ebroto** [comment]
>Right now, the rustfmt component is missing from nightly and it seems that when building clippy with a nightly that does not offer a required (in the rust-toolchain file) component, rustup simply skips the component.

If I'm not mistaken the situation should not be different to what we have now as we are using `rustfmt` from the `nightly` (without a specific date) toolchain for the tests.

> It also installs clippy from that nightly for some reason which seems a bit excessive? smile

Yeah, that was an oversight when adding TOML support for the toolchain file, it install the components in the current profile in addition to the ones specified. There's rust-lang/rustup#2579 to address it.

In the CI we set the minimal profile before starting the build, so it does not try to install clippy.

File: doc/basics.md
Comment (flip1995) [reply doc/basics.md:41]: They need to run `rustup self update` at least. `cargo` will download the toolchain specified in the `rust-toolchain` file automatically though.

**Comment by ebroto** [comment]
After testing this in the in-tree version of Clippy, I *think* it does not cause additional problems. I ran into the `Found multiple rlibs for crate serde` error, but I think it's not related to this change.

I've opened a [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Found.20multiple.20rlibs.20for.20crate.20serde/near/218234321) to discuss that problem, but I think it was already present.

**Comment by Manishearth** [comment]
I am a bit iffy about this: we _used to_ pin to nightlies and dealing with the gap period between master merges and nightlies was really annoying. But now that rustups can happen without us needing to fix things, perhaps this just means we need to run syncs at appropriate times. That seems fine.

**Comment by flip1995** [comment]
>  But now that rustups can happen without us needing to fix things, perhaps this just means we need to run syncs at appropriate times. That seems fine.

Yeah, now it should just be `git sbtree push` and `git mergetool`. We then have to figure out the right cycle, so that the `mergetool` won't be overwhelmed. 

I'll try to review this over the weekend :+1:

**Comment by flip1995** [comment]
Hijacking this PR to re-open the tree:

@bors treeclosed-

**Review by flip1995 — APPROVED** [review]
LGTM, let's do this with https://github.com/rust-lang/rust-clippy/pull/6401#discussion_r532261734 addressed 🚀 

Also the sync should be in the order

1. `subtree push` to the Clippy repo
2. `subtree pull` to the Rust repo

That way we make sure, that both Clippy and Rust pass the complete Clippy test suite after each sync.

Also somewhere should be documented, that you want to sync from Rust to Clippy from the latest nightly commit (`rustup check`), not the top-of-tree commit.

**Comment by ebroto** [comment]
I've addressed the last remark and updated to the latest nightly to be coherent with the latest rustup we did yesterday.

**Comment by flip1995** [comment]
@bors r+

Let's go, thanks for moving this forward! :rocket: 

Next steps:

1. Adapt `.rustfmt.toml` from rustc
2. Try to use the same nightly `rustfmt` as rustc
3. Let the `tidy` check in rustc also check Clippy.

**Comment by flip1995** [comment]
Off to a godd start :smile: 

I don't think we need the steps, which set the `LD_LIBRARY_PATH` and equivalents anymore.

**Comment by flip1995** [comment]
This patch should do the job:

```diff
diff --git a/.github/workflows/clippy.yml b/.github/workflows/clippy.yml
index 181b3bab4..7a9a2893c 100644
--- a/.github/workflows/clippy.yml
+++ b/.github/workflows/clippy.yml
@@ -38,7 +38,6 @@ jobs:
     - name: rust-toolchain
       uses: actions-rs/toolchain@v1.0.6
       with:
-        toolchain: nightly
         target: x86_64-unknown-linux-gnu
         profile: minimal

@@ -56,12 +55,6 @@ jobs:
         restore-keys: |
           ${{ runner.os }}-x86_64-unknown-linux-gnu

-    # Run
-    - name: Set LD_LIBRARY_PATH (Linux)
-      run: |
-        SYSROOT=$(rustc --print sysroot)
-        echo "LD_LIBRARY_PATH=${SYSROOT}/lib${LD_LIBRARY_PATH+:${LD_LIBRARY_PATH}}" >> $GITHUB_ENV
-
     - name: Build
       run: cargo build --features deny-warnings

diff --git a/.github/workflows/clippy_bors.yml b/.github/workflows/clippy_bors.yml
index f08182365..06fb314d1 100644
--- a/.github/workflows/clippy_bors.yml
+++ b/.github/workflows/clippy_bors.yml
@@ -87,7 +87,6 @@ jobs:
     - name: rust-toolchain
       uses: actions-rs/toolchain@v1.0.6
       with:
-        toolchain: nightly
         target: ${{ matrix.host }}
         profile: minimal

@@ -106,23 +105,6 @@ jobs:
           ${{ runner.os }}-${{ matrix.host }}

     # Run
-    - name: Set LD_LIBRARY_PATH (Linux)
-      if: runner.os == 'Linux'
-      run: |
-        SYSROOT=$(rustc --print sysroot)
-        echo "LD_LIBRARY_PATH=${SYSROOT}/lib${LD_LIBRARY_PATH+:${LD_LIBRARY_PATH}}" >> $GITHUB_ENV
-    - name: Link rustc dylib (MacOS)
-      if: runner.os == 'macOS'
-      run: |
-        SYSROOT=$(rustc --print sysroot)
-        sudo mkdir -p /usr/local/lib
-        sudo find "${SYSROOT}/lib" -maxdepth 1 -name '*dylib' -exec ln -s {} /usr/local/lib \;
-    - name: Set PATH (Windows)
-      if: runner.os == 'Windows'
-      run: |
-        SYSROOT=$(rustc --print sysroot)
-        echo "$SYSROOT/bin" >> $GITHUB_PATH
-
     - name: Build with internal lints
       run: cargo build --features deny-warnings,internal-lints

@@ -169,7 +151,6 @@ jobs:
     - name: rust-toolchain
       uses: actions-rs/toolchain@v1.0.6
       with:
-        toolchain: nightly
         target: x86_64-unknown-linux-gnu
         profile: minimal

@@ -247,7 +228,6 @@ jobs:
     - name: rust-toolchain
       uses: actions-rs/toolchain@v1.0.6
       with:
-        toolchain: nightly
         target: x86_64-unknown-linux-gnu
         profile: minimal

diff --git a/.github/workflows/clippy_dev.yml b/.github/workflows/clippy_dev.yml
index 5ee157cf2..8ca6a8c0c 100644
--- a/.github/workflows/clippy_dev.yml
+++ b/.github/workflows/clippy_dev.yml
@@ -25,10 +25,8 @@ jobs:
     - name: rust-toolchain
       uses: actions-rs/toolchain@v1.0.6
       with:
-        toolchain: nightly
         target: x86_64-unknown-linux-gnu
         profile: minimal
-        components: rustfmt

     - name: Checkout
       uses: actions/checkout@v2.3.3
```

**Comment by flip1995** [comment]
^We also have to remove the `toolchain: nightly`, because otherwise it installs the latest `nightly` and on `cargo build` it will install the specified nightly, according to the [actions-rs/toolchain documentation](https://github.com/actions-rs/toolchain#the-toolchain-file).

**Comment by ebroto** [comment]
>^We also have to remove the toolchain: nightly, because otherwise it installs the latest nightly and on cargo build it will install the specified nightly, according to the actions-rs/toolchain documentation.

Hmm don't we need `nightly` for the fmt tests?

**Comment by ebroto** [comment]
>Error: toolchain input was not given and repository does not have a rust-toolchain file

Oh right, we need something to start with, because the toolchain file will only be read after `cargo` etc have been installed for some toolchain

**Comment by ebroto** [comment]
Sorry for the spam!

So, I think we can remove nightly for all builds except clippy_dev for the fmt test, if we move the checkout step before the toolchain step so the toolchain file is found.

>++ ./target/debug/clippy-driver --print sysroot
./target/debug/clippy-driver: error while loading shared libraries: librustc_driver-74194b4af230f3a4.so: cannot open shared object file: No such file or directory
>+ sysroot=
Error: Process completed with exit code 127.

For the LD_LIBRARY_PATH stuff, I will take a look later

**Comment by flip1995** [comment]
> So, I think we can remove nightly for all builds except clippy_dev for the fmt test, if we move the checkout step before the toolchain step so the toolchain file is found.

Oh yeah, your right, we should keep it for the `clippy_dev` tests. 

> Oh right, we need something to start with, because the toolchain file will only be read after `cargo` etc have been installed for some toolchain

This happens because the `rust-toolchain` step is before the `checkout` step. Which was intentional before (for some reason I don't remember) but now we want to install the toolchain after the `checkout` step.

**Comment by ebroto** [comment]
I think the initial error we saw is due to `actions-rs/toolchain` assuming the old format for the toolchain file
```
 /usr/share/rust/.cargo/bin/rustup toolchain install [toolchain]
...
 error: invalid toolchain name: '[toolchain]
```
I'm looking to see what I can do, not sure why was not triggering in the PR build though :thinking: 

---

> Working on the CI without spamming, isn't really working on CI, so keep spamming :)

I managed to trigger the bors build manually in my fork so luckily I can avoid it :)

**Comment by flip1995** [comment]
@ebroto I had some time to look at this myself and pushed my fix for this directly to your branch. Everything should work now, but I'd like for you to sanity check my changes.

@bors try

**Comment by flip1995** [comment]
@bors try

Let's try to enable all of our integration tests again

**Comment by flip1995** [comment]
Uhm some changes in this PR make our integration test silently fail/not run at all. I'll investigate.

File: .github/workflows/clippy.yml
Comment (ebroto) [inline .github/workflows/clippy.yml:42]: If we remove the `rust-toolchain` step, from where does the `rustup` used in this command come from?

**Review by ebroto — COMMENTED** [review]
LGTM in general (github won't let me auto-approve the PR :smile:)

I just left a doubt.

File: .github/workflows/clippy.yml
Comment (flip1995) [reply .github/workflows/clippy.yml:42]: GHA ships with stable rust+rustup already installed. See [GHA documentation about the pre-installed software](https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners#supported-software)

Comment (ebroto) [reply .github/workflows/clippy.yml:42]: That's convenient :)

BTW thanks for the work on this, you saved me another night fighting the CI :smile:

Comment (flip1995) [reply .github/workflows/clippy.yml:42]: NP, I like CI pain :smile: 

So can we merge this? :)

Comment (ebroto) [reply .github/workflows/clippy.yml:42]: Sure! I was hesitant about approving a PR I opened :smile:

**Pull request merged**
